### ISSUE TITLE
Pagination

### ISSUE THREAD

**Issue by SoloVeniaASaludar** [ocpi/ocpi#373]
The specification says

```
Some query parameters can cause concurrency problems. For example the date_to query parameter. When
there are for example 1000 objects matching a query for all objects with date_to before 2016-01-01. While
crawling over the pages one of these objects is updated. The client detects this: X-Total-Count will be lower in
the next request. It is advised to redo the previous GET with the offset lowered by 1 (if the offset was not 0)
and after that continue crawling the 'next' page links.

```

By example, in the case of pagination for a GET to the Locations sender interface, assume the sender has objects with the following last_update times:

2021-01-01 01:00:00
2021-01-01 02:00:00
2021-01-01 03:00:00
2021-01-01 04:00:00
2021-01-01 05:00:00
2021-01-01 06:00:00
2021-01-01 07:00:00
2021-01-01 08:00:00
2021-01-01 09:00:00

The receiver sends a GET to the sender with date_from=2021-01-01 00:00:00, limit=3, offset=0. The answer will contain the data for:
2021-01-01 01:00:00
2021-01-01 02:00:00
2021-01-01 03:00:00

Now, assume location with timestamp 2021-01-01 02:00:00 is updated, its time being now 2021-01-02 02:00:00. That means the full list at sender is now:

2021-01-01 01:00:00
2021-01-01 03:00:00
2021-01-01 04:00:00
2021-01-01 05:00:00
2021-01-01 06:00:00
2021-01-01 07:00:00
2021-01-01 08:00:00
2021-01-01 09:00:00
2021-01-02 02:00:00

When the receiver sends next GET with date_from=2021-01-01 01:00:00, limit=3, offset=3. It will receive:
2021-01-01 05:00:00
2021-01-01 06:00:00
2021-01-01 07:00:00

that is an error because "2021-01-01 04:00:00" has been forgotten. 

Thus, loop control via "offset" has dangerous scenarios, being this one an example. Some other problematic use cases can be found that results in forgotten or repeated data: an already reported item can change status, ... .

I'm newcormer in OCPI, learning it, some doubts about pagination appear:

1) we need pagination ?

Assume the receiver (after some time out-of-service) wants update its data. It sends a GET to the receiver with "date_from=XXX" (start of out-of-service). The receiver answers ONLY with "200 OK", no body, and starts a loop of POST/PUT to the receiver sending, one by one, the requested objects. Do not send a POST&PUT until answer to previous one is received and status is 200 OK (a status different of 200 OK allows the receiver to interrupt the loop). No need for limit, offset, additional headers, ... .

 2) if pagination is necessary, must the sender or the receiver control it ?

Assume pagination is fully controlled by the sender. The receiver starts sending a GET with some limit and receives an answer with first objects and a "next" link. Receiver queries the next link, receiving more objects and another "next" link. Receiver repeats until no "next" link is received. In this way, the loop is fully controlled by the sender. No need for offset, total_count, ... .

**Comment by reinierl** [comment]
Yes, pagination in OCPI 2.x has issues and we're aware of the solutions you are proposing. OCPI 2.2.1 is what it is for now and we're not planning on making major changes to it before OCPI 3.0.

Who are you working for if I may ask? If they are or are willing to become a member of the EV Roaming Association we would be happy to get your feedback directly in the OCPI development process.

**Comment by reinierl** [comment]
Or well, we we were aware of your proposals under questions 2 and 3 but your idea of "requesting a push" under (1) I hadn't considered and it seems elegant too. It kind of reminds of how Firebase works where the receiver subscribes once and then gets pushed historic changes and real-time changes in one stream.
